RavenDB-12771 fixing resolved document and revisions behaviour

karmeli87 opened the pull request:
- modifing a resovled document should keep the HasRevisions flags
- A revision that has the 'Conflicted' flag shouldn't overwrite the document
- fix various database change vector merging
- adding counter to a resovled document should not generate a revision

karmeli87 committed:
RavenDB-12771 fixing resolved document and revisions behaviour

- modifing a resovled document should keep the HasRevisions flags
- A revision that has the 'Conflicted' flag shouldn't overwrite the document
- fix various database change vector merging
- adding counter to a resovled document should not generate a revision

grisha-kotler committed:
RavenDB-12771 - fix failing test

grisha-kotler committed:
RavenDB-12771 - delete all revisions from storage when MinimumRevisionsToKeep = 0

ayende commented on src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:
```diff
@@ -250,9 +250,11 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
             return Configuration.Default ?? _emptyConfiguration;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
This is a big method and not likely to be inlined even with this.
Why is this here?

grisha-kotler replied:
done

ayende commented on src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:
```diff
@@ -312,6 +321,7 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
remove this

grisha-kotler replied:
done

ayende reviewed (commented):
Talk to me about this tomorrow, I want to go over this with you so I can be sure that I understand what you have done here

karmeli87 committed:
RavenDB-12771 fix failing tests and logic

- Strip the 'Resolved' flag from a document that is updated via the counter.
- Resolved document revision will kept with the merged change vector, but the actuall one we generate a new change vector so it can be replicated. (needed for external replication)

grisha-kotler committed:
RavenDB-12771 - PR comments

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -693,7 +693,7 @@ public long GetNumberOfCounterEntries(DocumentsOperationContext context)
                 return;
 
-            var flags = doc.Flags.Strip(DocumentFlags.FromClusterTransaction);
+            var flags = doc.Flags.Strip(DocumentFlags.FromClusterTransaction | DocumentFlags.Resolved);
```
Check what happens with attachments in the same scenario

karmeli87 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -229,6 +222,20 @@ oldFlags .Contain(DocumentFlags.HasAttachments))
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Remove this

karmeli87 replied:
done

karmeli87 committed:
RavenDB-12771 address PR comments

grisha-kotler committed:
RavenDB-12771 - add a test to verify that we deleted all revisions when MinimumRevisionsToKeep = 0
